packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts

```diff
@@ -1482,68 +2054,86 @@ interface CrossFieldTable<TFieldData> {
-	 *
-	 * This map is needed once we switch from the initial pass (which generates a new changeset) to the second pass which
-	 * performs surgery on the changeset generated in the first pass: we don't want to re-run the rebasing of nested
-	 * changes. Instead we want to keep using the objects generated in the first pass and mutate them where needed.
+	 * Maps from the FieldChange key used for the CrossFieldTable (which is the base FieldChange)
+	 * to the context for the field.
 	 */
-	rebasedNodeCache: Map<NodeChangeset, NodeChangeset>;
+	readonly baseFieldToContext: Map<FieldChange, RebaseFieldContext>;
+	readonly baseToRebasedNodeId: ChangeAtomIdMap<NodeId>;
+	readonly rebasedFields: Set<FieldChange>;
+	readonly rebasedNodeToParent: ChangeAtomIdMap<FieldId>;
+	readonly rebasedCrossFieldKeys: CrossFieldKeyTable;
 
 	/**
 	 * List of (newId, baseId) pairs encountered so far.
 	 */
-	nodeIdPairs: [NodeId | undefined, NodeId | undefined][];
+	readonly nodeIdPairs: [NodeId, NodeId, NodeAttachState | undefined][];
+
+	readonly affectedNewFields: TupleBTree<FieldIdKey, boolean>;
+	readonly affectedBaseFields: TupleBTree<FieldIdKey, boolean>;
 }
 
+type FieldIdKey = [RevisionTag | undefined, ChangesetLocalId | undefined, FieldKey];
+
 interface RebaseFieldContext {
 	baseChange: FieldChange;
 	newChange: FieldChange;
 	rebasedChange: FieldChange;
+	fieldId: FieldId;
+
+	/**
+	 * The set of node IDs in the base changeset which should be included in the rebased field,
+	 * even if there is no corresponding node changeset in the new change.
+	 */
+	baseNodeIds: NodeId[];
 }
 
-function newComposeTable(): ComposeTable {
+function newComposeTable(
+	baseChange: ModularChangeset,
+	newChange: ModularChangeset,
+): ComposeTable {
 	return {
 		...newCrossFieldTable<FieldChange>(),
+		baseChange,
+		newChange,
 		fieldToContext: new Map(),
-		nodeIds: new Map(),
+		newFieldToBaseField: new Map(),
+		newToBaseNodeId: new Map(),
 		nodeIdPairs: [],
+		composedNodes: new Set(),
+		affectedBaseFields: newBTree(),
+		affectedNewFields: newBTree(),
 	};
 }
 
 interface ComposeTable extends CrossFieldTable<FieldChange> {
-	/**
-	 * Maps from an input changeset for a field (from change1 if it has one, from change2 otherwise) to the context for that field.
-	 */
-	fieldToContext: Map<FieldChange, ComposeFieldContext>;
+	readonly baseChange: ModularChangeset;
+	readonly newChange: ModularChangeset;
 
 	/**
-	 * The set of node IDs from the second changeset which have been encountered.
+	 * Maps from an input changeset for a field (from change1 if it has one, from change2 otherwise) to the context for that field.
 	 */
-	nodeIds: NestedSet<RevisionTag, ChangesetLocalId>;
-
-	nodeIdPairs: [NodeId | undefined, NodeId | undefined][];
+	readonly fieldToContext: Map<FieldChange, ComposeFieldContext>;
+	readonly newFieldToBaseField: Map<FieldChange, FieldChange>;
+
+	// TODO: This could just be new IDs
+	readonly nodeIdPairs: [NodeId, NodeId][];
```

Comment:
Would be good to have a doc comment explaining that each pair represents the base and new changes to a given node, and explain the lifetime of this information as part of the compostion process.